cpp/src/arrow/compute/exec/expression_test.cc

```diff
@@ -69,6 +70,39 @@ Expression true_unless_null(Expression argument) {
   return call("true_unless_null", {std::move(argument)});
 }
 
+Expression add(Expression l, Expression r) {
+  return call("add", {std::move(l), std::move(r)});
+}
+
+namespace arrow_literals {
+
+using namespace std::chrono_literals;
+
+inline auto operator""_ts(const char* c, size_t s) {
+  return [s = StringScalar(std::string{c, s})](TimeUnit::type unit) {
+    return literal(*s.CastTo(timestamp(unit)));
+  };
+}
+
+}  // namespace arrow_literals
+
+template <typename Rep, typename Period>
+Expression literal(std::chrono::duration<Rep, Period> d) {
+  int64_t int_value = d.count();
+  TimeUnit::type unit;
+  if constexpr (std::is_same_v<Period, std::nano>) {
+    unit = TimeUnit::NANO;
+  } else if constexpr (std::is_same_v<Period, std::micro>) {
+    unit = TimeUnit::MICRO;
+  } else if constexpr (std::is_same_v<Period, std::milli>) {
+    unit = TimeUnit::MILLI;
+  } else {
+    unit = TimeUnit::SECOND;
+    int_value = std::chrono::duration_cast<std::chrono::seconds>(d).count();
+  }
+  return literal(*MakeScalar(int_value)->CastTo(duration(unit)));
+}
```

Comment:
nit: This seems generally useful (note, I thought that on the earlier version before the literals route too) to be something hidden in a test case.  Maybe a follow-up PR to add tests and expose the functionality at some point (though it's probably just one in a sea of QoL features we could offer C++ users).

Reply:
The logical public place to include it would be as a constructor for `Datum`, but that'd incur `#include <chrono>` for everything which uses Datum. Maybe that's an acceptable cost, WDYT @pitrou ?

Reply:
How large is `<chrono>` typically? Perhaps we can add a lower-level helper to `arrow/util/time.h`:
```c++
template <typename Rep, typename Period>
int64_t DurationToUnit(std::chrono::duration<Rep, Period> d, TimeUnit::type unit) {
  ...
}
```